lib/logitech_receiver/common.py

```diff
@@ -546,18 +551,57 @@ def __getattr__(self, k):
 """Firmware information."""
 FirmwareInfo = namedtuple("FirmwareInfo", ["kind", "name", "version", "extras"])
 
-BATTERY_APPROX = NamedInts(empty=0, critical=5, low=20, good=50, full=90)
-
-BATTERY_STATUS = NamedInts(
-    discharging=0x00,
-    recharging=0x01,
-    almost_full=0x02,
-    full=0x03,
-    slow_recharge=0x04,
-    invalid_battery=0x05,
-    thermal_error=0x06,
-)
 
-
-def BATTERY_OK(status):
-    return status not in (BATTERY_STATUS.invalid_battery, BATTERY_STATUS.thermal_error)
+@dataclass
+class Battery:
+    """Information about the current state of a battery"""
+
+    level: Optional[NamedInt | int]
+    next_level: Optional[NamedInt | int]
+    status: Optional[NamedInt]
+    voltage: Optional[int]
+    light_level: Optional[int] = None  # light level for devices with solaar recharging
+
+    def __post_init__(self):
+        if self.level is None:  # infer level from status if needed and possible
+            if self.status == Battery.STATUS.full:
+                self.level = Battery.APPROX.full
+            elif self.status in (Battery.STATUS.almost_full, Battery.STATUS.recharging):
+                self.level = Battery.APPROX.good
+            elif self.status == Battery.STATUS.slow_recharge:
+                self.level = Battery.APPROX.low
+
+    STATUS = NamedInts(
+        discharging=0x00,
+        recharging=0x01,
+        almost_full=0x02,
+        full=0x03,
+        slow_recharge=0x04,
+        invalid_battery=0x05,
+        thermal_error=0x06,
+    )
+
+    APPROX = NamedInts(empty=0, critical=5, low=20, good=50, full=90)
+
+    ATTENTION_LEVEL = 5
+
+    def ok(self):
+        return self.status not in (Battery.STATUS.invalid_battery, Battery.STATUS.thermal_error) and (
+            self.level is None or self.level > Battery.ATTENTION_LEVEL
+        )
+
+    def charging(self):
+        return self.status in (
+            Battery.STATUS.recharging,
+            Battery.STATUS.almost_full,
+            Battery.STATUS.full,
+            Battery.STATUS.slow_recharge,
+        )
+
+    def to_str(self):
+        if isinstance(self.level, NamedInt):
+            return _("Battery: %(level)s (%(status)s)") % {"level": _(self.level), "status": _(self.status)}
```

Comment:
Much, much shorter with f-strings.

return _("Battery: {_(self.level)} ({_(self.status)})")

Reply:
There needs to be an f in there somewhere.  And then how does translation work on f-strings?